test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -266,26 +279,117 @@ func responsesConfigMap(name string, namespace *corev1.Namespace) *corev1.Config
+
+			// Non-blocking hooks are set to Status:Success.
+			"AfterControlPlaneInitialized-preloadedResponse": `{"Status": "Success"}`,
+			"AfterClusterUpgrade-preloadedResponse":          `{"Status": "Success"}`,
 		},
 	}
 }
 
+// Check that each hook in hooks has been called at least once by checking if its actualResponseStatus is in the hook response configmap.
+// If the provided hooks have both keys and values check that the values match those in the hook response configmap.
 func checkLifecycleHooks(ctx context.Context, c client.Client, namespace string, clusterName string, hooks map[string]string) error {
 	configMap := &corev1.ConfigMap{}
 	configMapName := clusterName + "-hookresponses"
 	err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: configMapName}, configMap)
 	Expect(err).ToNot(HaveOccurred(), "Failed to get the hook response configmap")
-	for hook := range hooks {
-		if _, ok := configMap.Data[hook+"-called"]; !ok {
+	for hook, expected := range hooks {
+		v, ok := configMap.Data[hook+"-actualResponseStatus"]
+		if !ok {
 			return errors.Errorf("hook %s call not recorded in configMap %s/%s", hook, namespace, configMapName)
 		}
+		if expected != "" && expected != v {
+			return errors.Errorf("hook %s was expected to be %s in configMap got %s", expected, hook, v)
+		}
 	}
 	return nil
 }
+
+// beforeClusterCreateTestHandler provides an unblocked function which returns true if the Cluster has entered ClusterPhaseProvisioned.
+func beforeClusterCreateTestHandler(ctx context.Context, c client.Client, namespace, clusterName string, intervals []interface{}) {
+	runtimeHookTestHandler(ctx, c, namespace, clusterName, "BeforeClusterCreate", func() bool {
+		// This hook should block the Cluster from entering the "Provisioned" state.
+		cluster := &clusterv1.Cluster{}
+		Expect(c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, cluster)).To(Succeed())
+		return cluster.Status.Phase == string(clusterv1.ClusterPhaseProvisioned)
+	}, intervals)
+}
+
+// beforeClusterUpgradeTestHandler provides an unblocked function which returns true if the Cluster has controlplanev1.RollingUpdateInProgressReason in its
+// ReadyCondition.
+func beforeClusterUpgradeTestHandler(ctx context.Context, c client.Client, namespace, clusterName, version string, intervals []interface{}) {
+	runtimeHookTestHandler(ctx, c, namespace, clusterName, "BeforeClusterUpgrade", func() bool {
+		cluster := &clusterv1.Cluster{}
+		var unblocked bool
+
+		// First ensure the Cluster topology has been updated to the target Kubernetes Version.
+		Eventually(func() bool {
+			Expect(c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, cluster)).To(Succeed())
+			return cluster.Spec.Topology.Version == version
+		}).Should(BeTrue(), "BeforeClusterUpgrade blocking condition false: Cluster topology has not been updated to the target Kubernetes Version")
```

Comment:
```suggestion
		}).Should(BeTrue(), "Cluster topology reconciliation failed: Cluster topology has not been updated to the target Kubernetes Version")
```
nit: I think it's a bit confusing given that that's a failure of the topology reconciler we detect here. And the blocking condition is or won't be really false this Eventually will just fail the entire test directly

Do we need a timeout here? I wonder if it's racy otherwise if the topology reconciler doesn't act fast enough